File: taller_hilos/ejercicio2/Ejercicio2.java
// Ejercicio 2: Dos hilos que imprimen mensajes diferentes
//
// Cada clase representa una tarea independiente. Al lanzar ambos hilos
// con start(), el sistema operativo los intercala según su planificador,
// por eso el orden de los mensajes puede variar en cada ejecución.

class MensajeA implements Runnable {
    public void run() {
        // Hilo A imprime su propio mensaje 5 veces.
        for (int i = 0; i < 5; i++) {
            System.out.println("Hilo A: Hola desde el primer hilo");
        }
    }
}

class MensajeB implements Runnable {
    public void run() {
        // Hilo B imprime un mensaje distinto al de A, también 5 veces.
        // Ambos hilos corren al mismo tiempo, así que sus líneas
        // pueden aparecer entremezcladas en la consola.
        for (int i = 0; i < 5; i++) {
            System.out.println("Hilo B: Hola desde el segundo hilo");
        }
    }
}

public class Ejercicio2 {
    public static void main(String[] args) {
        Thread hiloA = new Thread(new MensajeA());
        Thread hiloB = new Thread(new MensajeB());

        // Ambos start() se llaman uno tras otro en main.
        // Java lanza los dos hilos casi simultáneamente;
        // no esperamos a que A termine para arrancar B.
        hiloA.start();
        hiloB.start();
    }
}

File: taller_hilos/ejercicio2/ejercicio2.py
import threading

# Ejercicio 2: Dos hilos que imprimen mensajes diferentes
#
# Al lanzar dos hilos sin sincronización, el GIL de Python los alterna
# automáticamente. El orden exacto de impresión no está garantizado.

def mensaje_a():
    for _ in range(5):
        print("Hilo A: Hola desde el primer hilo")

def mensaje_b():
    for _ in range(5):
        print("Hilo B: Hola desde el segundo hilo")

# Creamos un hilo por cada función-tarea.
hilo_a = threading.Thread(target=mensaje_a)
hilo_b = threading.Thread(target=mensaje_b)

# Arrancamos ambos hilos. A partir de aquí corren concurrentemente.
hilo_a.start()
hilo_b.start()

# Esperamos a que los dos terminen antes de que el programa finalice.
hilo_a.join()
hilo_b.join()
